src/coreclr/jit/lower.cpp

```diff
@@ -7844,6 +7860,126 @@ void Lowering::ContainCheckBitCast(GenTree* node)
     }
 }
 
+//------------------------------------------------------------------------
+// LowerBlockStoreAsHelperCall: Lower a block store node as a memset/memcpy call
+//
+// Arguments:
+//    blkNode - The block store node to lower
+//
+void Lowering::LowerBlockStoreAsHelperCall(GenTreeBlk* blkNode)
+{
+    // We shouldn't be using helper calls for blocks on heap containing GC pointers.
+    // due to atomicity guarantees.
+    assert(!blkNode->IsZeroingGcPointersOnHeap());
+
+    LIR::Use use;
+    assert(!BlockRange().TryGetUse(blkNode, &use));
+
+    const bool isVolatile = blkNode->IsVolatile();
+
+    GenTree* dest = blkNode->Addr();
+    GenTree* data = blkNode->Data();
+    GenTree* size;
+
+    // Is it Memset ...
+    if (blkNode->OperIsInitBlkOp())
+    {
+        // Drop GT_INIT_VAL nodes
+        if (data->OperIsInitVal())
+        {
+            BlockRange().Remove(data);
+            data = data->gtGetOp1();
+        }
+    }
+    else
+    {
+        // ... or Memcpy?
+        assert(data->OperIs(GT_IND, GT_LCL_VAR, GT_LCL_FLD));
+
+        // Drop GT_IND nodes
+        if (data->OperIs(GT_IND))
+        {
+            BlockRange().Remove(data);
+            data = data->AsIndir()->Addr();
+        }
```

Comment:
This doesn't handle the `GT_LCL_VAR` or `GT_LCL_FLD` cases correctly (would need to convert them to `GT_LCL_ADDR`). Do we ever see them here?

Reply:
Ah sorry, I somehow missed that. Yeah seems like it could be a bit simpler (also maybe make it an `else`?)

Reply:
Addressed!